compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs

```diff
@@ -389,18 +389,13 @@ pub(super) fn find_opaque_ty_constraints_for_rpit<'tcx>(
             // the `concrete_opaque_types` table.
             Ty::new_error(tcx, guar)
         } else {
-            // Fall back to the RPIT we inferred during HIR typeck
-            if let Some(hir_opaque_ty) = hir_opaque_ty {
-                hir_opaque_ty.ty
-            } else {
-                // We failed to resolve the opaque type or it
-                // resolves to itself. We interpret this as the
-                // no values of the hidden type ever being constructed,
-                // so we can just make the hidden type be `!`.
-                // For backwards compatibility reasons, we fall back to
-                // `()` until we the diverging default is changed.
-                Ty::new_diverging_default(tcx)
-            }
+            // We failed to resolve the opaque type or it
+            // resolves to itself. We interpret this as the
+            // no values of the hidden type ever being constructed,
+            // so we can just make the hidden type be `!`.
+            // For backwards compatibility reasons, we fall back to
+            // `()` until we the diverging default is changed.
+            Ty::new_diverging_default(tcx)
```

Comment:
it feels like the PR title isn't really accurate: you fall back to `()` in case an RPIT is not defined by the MIR. I expect the following example to therefore compile (but with different behavior), which seems very undesirable to me
```rust
use std::fmt::Debug;
fn foo() -> impl Debug + Default {
    panic!();
    1u32
}

fn print_ret<F: FnOnce() -> R, R: Debug + Default>(_: F) {
    println!("{:?}", R::default());
}

fn main() {
    print_ret(foo);
}

```

Reply:
Ohh lol, my mistake. That explains why the weird behavior.
Going to fix this later.